### ISSUE TITLE
Boost.JSON output for numbers with scientific format is incorrect

### ISSUE THREAD

**Issue by sicheste** [miloyip/nativejson-benchmark#156]
When enabling in main.cpp lines 1139 and 1140 it is showing the differences in parsing. For the Boost.JSON it will for example print this:
```
Expect: [5e-324]
Actual: [5E-324]
```
Intput from roundtrip24.json is simply `[5e-324]`.

Somewhere the formatting is broken. By default Boost.JSON would output `e` instead of `E`:

```cpp
#include <boost/json/src.hpp>
#include <iostream>
#include <string>

int main(int, char**) {
  std::string json("[5e-342]");
  std::string result = boost::json::serialize(json);
  std::cout << "input: " << json << "\nresult: " << result << std::endl;
  return 0;
}
```
Output:
```
input: [5e-342]
result: "[5e-342]"
```
The exponent e is lower case here. I did not had enough time to search for some formatting settings in the application.

**Comment by niXman** [comment]
@sicheste 

> Expect: [5e-324]
> Actual: [5E-324]

is there any reason except aesthetic?

and why you are posting it here instead of the boost.json developer project?

**Comment by sicheste** [comment]
> is there any reason except aesthetic?

Yes, the test application is treating this as an invalid result. It is not making a string to lower or similar before comparing the result to the expected value.


> and why you are posting it here instead of the boost.json developer project?

Because this formatting is set somewhere within the test application. As you can se in my minimal example from above, it is not caused by Boost.JSON. Maybe it is set by another JSON library. I do not know why, I have not digged deeper. But it is not the default behaviour of Boost.JSON.

**Comment by niXman** [comment]
> When enabling in main.cpp

where is that file?

**Comment by sicheste** [comment]
> > When enabling in main.cpp
> 
> where is that file?

https://github.com/miloyip/nativejson-benchmark/blob/master/src/main.cpp#L1139

**Comment by niXman** [comment]
but according to the JSON grammar: https://www.json.org/json-en.html, the exponent char may be one of `e` or `E`. so from a formal point of view everything is correct here.

**Comment by sicheste** [comment]
But then this means that the comparision within the benchmark application is incorrect. So Boost.JSON is blamed to be not be fully following the standard by mistake.

I am totally fine to rephrase the title of this issue. But currently Boost.JSON is giving a result which seems correct, the test application does not accept it as being correct. When, regarding to the JSON specificition, it is fine to use `e` or `E`, then both should be accepted.

**Comment by niXman** [comment]
first - according to the [doc](https://www.boost.org/doc/libs/1_81_0/libs/json/doc/html/json/ref/boost__json__serialize.html) it seems your example uses `boost::json::serialize(string)` which [call](https://github.com/boostorg/json/blob/master/include/boost/json/impl/serialize.ipp#L97) for [serialize_impl()](https://github.com/boostorg/json/blob/master/include/boost/json/impl/serialize.ipp#L103), [reset(string *)](https://github.com/boostorg/json/blob/master/include/boost/json/impl/serializer.ipp#L784) and finally uses [write_string()](https://github.com/boostorg/json/blob/master/include/boost/json/impl/serializer.ipp#L211) which do nothing with doubles except for escaping the string.

second - try to change your code that it parse the string first and serialize it back as string. and then show the result.

**Comment by niXman** [comment]
the `E` comes from here: https://github.com/boostorg/json/blob/master/include/boost/json/detail/ryu/impl/d2s.ipp#L612
